### ISSUE TITLE
Input checking with preserved pandas dataframes

### ISSUE THREAD

**Issue by peterzsohar** [dask/dask-ml#584]
The current implementation of `check_array` with `preserve_pandas_dataframe=True` is only a pass-through.
https://github.com/dask/dask-ml/blob/7793962d990a4b5290a5cfd1aee1d5b48828f1a1/dask_ml/utils.py#L167
It's clear from the comment (`# TODO: validation?`) that there is (or at least was) intent to do some form of validation but I could not find any trace of that in the open issues or PRs.

I would be happy to give it a try if this is of general interest.

My approach would be to write a wrapper around the scikit-learn `check_array` function which preserves the axis labels and dtypes where applicable.
(As of now) I don't really see real obstacles as pandas is more strict with the numpy compatible datatypes so preserving those should not be a big issue. On the other hand categoricals might need some tinkering, though.
Also not entirely sure about how to support `pandas.SparseArray`s but there is a potential workaround with converting to scipy coo format which avoids densifying.

Any thoughts?

**Comment by TomAugspurger** [comment]
Happy to have a discussion here, though I would start with a concrete problem we have with an estimator, and work back to implementing something in `check_array` that handles it. Do we have any issues that are caught later on in training that would / could be caught earlier by a check_array?

For dask Array, the checks are primarily around array / chunk shapes and ndims, which don't really apply to DataFrame.

**Comment by peterzsohar** [comment]
Thanks for getting back to me on this.

I would not frame it as an issue for fixing something that's wrong but more like an additional feature that wouldn't really affect the current dask transformers* but would definitely increase consistency in projects which mix several dask-ml components with custom transformers only for their pandas support.

In general I personally find the idea very useful that I can rely on the same input checks whether my data has labels or not.

*=Currently only `SimpleImputer` and `OneHotEncoder` use `preserve_pandas_dataframe=True`. The latter's type check (`"categorical"`) could be caught in `check_array`. More generally I could see some cases of potentially redundant checks in other transformer-specific `_check_array` functions (eg. explicitly converting to array-like which could already be done in `check_array`, converting to `float` after `check_array`, custom `_check_array` exists only to check something that could be checked via the sklearn version of `check_array`, etc.).

**Comment by TomAugspurger** [comment]
OK, I think I understand, but to make sure...
 
When you have something like

```python
data = np.array([[1, np.inf, np.nan]])
MyEstimator().fit(data)
```

the call to `check_array` will would catch the NaN / inf values and raise? But if used via dask-ML and `preserve_pandas_dataframe=True`, it doesn't raise, since we skip the checks?

At this point, my question becomes: can we push this upstream to scikit-learn? For estimators like OneHotEncoder, we rely on the pandas categorical dtype, so scikit-learn's behavior of `check_array` coercing to an ndarray doesn't work for us. My preference is that `check_array` knows how to handle pandas data structures natively.

The answer to "can this be done in scikit-learn?" may be "no". I'm not sure, but I recall `check_array` being used to both check the values and to coerce somewhat arbitrary input to an array. But I think it's worth discussing (if it hasn't been already).
